File: src/com/implimentz/fragments/FragmentHolder.java
package info.wikiroutes.android.fragments;

import android.content.ComponentCallbacks;
import android.content.Context;
import android.content.res.Configuration;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Created by Alexander Efremenkov.
 * Date: 07.01.16, 11:40
 * In Intellij IDEA 15.0.1 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */
public final class FragmentHolder {

    private static boolean registered = false;
    private static final Map<Integer, List<FragmentData>> stack = new ConcurrentHashMap<>();

    public static void init(final Context context) {
        if(registered) {
            return;
        }

        context.registerComponentCallbacks(callback);
        registered = true;
    }

    private static final ComponentCallbacks callback = new ComponentCallbacks() {
        @Override
        public void onConfigurationChanged(Configuration newConfig) {
            notifyConfigurationChanged(newConfig);
        }
        @Override
        public void onLowMemory() {
            notifyLowMemory();
        }
    };

    private static void notifyConfigurationChanged(final Configuration newConfig) {
        for (final List<FragmentData> list : stack.values()) {
            for (final FragmentData data : list) {
                final Fragment fragment = data.getFragment();
                fragment.setConfigurationChanged();
                if (fragment.isShowing()) {
                    fragment.onPause();
                    fragment.onConfigurationChanged(newConfig);
                }
            }
        }
    }

    private static void notifyLowMemory() {
        for (final Map.Entry<Integer, List<FragmentData>> entry : stack.entrySet()) {
            for (final FragmentData data : entry.getValue()) {
                final Fragment fragment = data.getFragment();
                if (fragment.isShowing()) {
                    fragment.onLowMemory();
                }
            }
        }
    }

    static void register(final int id) {
        if (stack.containsKey(id)) {
            return;
        }

        stack.put(id, new CopyOnWriteArrayList<>());
    }

    static void unregister(final int id) {
        if (!stack.containsKey(id)) {
            return;
        }

        stack.remove(id);
    }

    static List<FragmentData> getStackById(final int id) {
        if(!stack.containsKey(id)) {
            return new ArrayList<>();
        }

        return stack.get(id);
    }
    
}

File: src/main/java/com/implimentz/fragments/FragmentHolder.kt
@file:Suppress("unused")

package com.implimentz.fragments

import android.content.ComponentCallbacks
import android.content.Context
import android.content.res.Configuration
import java.io.Serializable
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Created by ironz.
 * Date: 07.01.16, 11:40
 * In Intellij IDEA 15.0.1 Ultimate
 * email: implimentz@gmail.com
 * twitter: iamironz
 */
object FragmentHolder {

    private var registered: Boolean = false
    private val stack: MutableMap<Int, MutableList<Fragment<out Serializable>>> = ConcurrentHashMap()
    
    fun init(context: Context) {

        if (registered) {
            return
        }

        context.registerComponentCallbacks(callback)
        registered = true
    }

    private val callback = object : ComponentCallbacks {
        override fun onConfigurationChanged(newConfig: Configuration) {
            notifyConfigurationChanged(newConfig)
        }
        override fun onLowMemory() {
            //not needed
        }
    }

    private fun notifyConfigurationChanged(newConfig: Configuration) {
        stack.values.forEach {
            setForAllConfigChanges(it)
            callAllForConfigurationChanged(it, newConfig)
        }
    }

    private fun setForAllConfigChanges(it: MutableList<Fragment<out Serializable>>) {
        it.forEach {
            it.configurationChanged = true
        }
    }

    private fun callAllForConfigurationChanged(it: MutableList<Fragment<out Serializable>>, newConfig: Configuration) {
        it.filter {
            it.finished.not()
        }.forEach {
            it.onConfigurationChanged(newConfig)
            it.onPause()
        }
    }

    internal fun register(id: Int) {
        if (stack.containsKey(id)) {
            return
        }

        stack.put(id, ArrayList())
    }

    internal fun unregister(id: Int) {
        if (!stack.containsKey(id)) {
            return
        }

        stack.remove(id)
    }

    internal fun getStackById(id: Int): MutableList<Fragment<out Serializable>> {
        return stack[id] ?: CopyOnWriteArrayList()
    }
}
